sql/opt/memo: TestStatsQuality failed

sql/opt/memo.TestStatsQuality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12901992?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12901992?buildTab=artifacts#/) on master @ [39a94462ca56caef9feb0a7d08ef99e43b01cdc6](https://github.com/cockroachdb/cockroach/commits/39a94462ca56caef9feb0a7d08ef99e43b01cdc6):


```
                    }
                ],
                "histo_col_type": "VARCHAR(101)",
                "histo_version": 1,
                "name": "__auto__",
                "null_count": 0,
                "row_count": 10000
            }
        ]';
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11185/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/memo/memo_test_/memo_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/memo/testdata/stats_quality/tpch/q13:4:
        import [1 args]
        <no input to command>
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11185/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/memo/memo_test_/memo_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/memo/testdata/stats_quality/tpch/q13:18:
         SELECT
            c_count, count(*) AS custdist
        FROM (
            SELECT
                c_custkey,
                count(o_orderkey)
            FROM
                customer LEFT OUTER JOIN orders ON
                    c_custkey = o_custkey
                    AND o_comment NOT LIKE '%special%requests%'
            GROUP BY
                c_custkey
            ) AS c_orders (c_custkey, c_count)
        GROUP BY
            c_count
        ORDER BY
            custdist DESC,
            c_count DESC;
        output didn't match expected:
        @@ -20,11 +20,11 @@
               │    ├── key: (1)
               │    ├── fd: (1)-->(22)
               │    ├── right-join (hash)
               │    │    ├── save-table-name: q13_right_join_4
               │    │    ├── columns: c_custkey:1(int!null) o_orderkey:11(int) o_custkey:12(int) o_comment:19(varchar)
        -      │    │    ├── stats: [rows=503988.2, distinct(1)=148813, null(1)=0, distinct(11)=317522, null(11)=0, distinct(12)=99620.1, null(12)=0, distinct(19)=317522, null(19)=0]
        +      │    │    ├── stats: [rows=503988.2, distinct(1)=148813, null(1)=0, distinct(11)=317522, null(11)=1.10795e-11, distinct(12)=99620.1, null(12)=0, distinct(19)=317522, null(19)=1.10795e-11]
               │    │    ├── key: (1,11)
               │    │    ├── fd: (11)-->(12,19)
               │    │    ├── select
               │    │    │    ├── save-table-name: q13_select_5
               │    │    │    ├── columns: o_orderkey:11(int!null) o_custkey:12(int!null) o_comment:19(varchar!null)
        --- FAIL: TestStatsQuality/tpch/q13 (0.15s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatsQuality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/opt/memo.TestStatsQuality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12903131?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12903131?buildTab=artifacts#/) on master @ [02e5b0ccffbbd9430d377fd9394c234cd73ad954](https://github.com/cockroachdb/cockroach/commits/02e5b0ccffbbd9430d377fd9394c234cd73ad954):


```
        column_names  row_count  distinct_count  null_count
        {o_d_id}      300000     10              0
        {o_id}        300000     2999            0
        {o_ol_cnt}    300000     11              0
        {o_w_id}      300000     10              0
        ~~~~
        column_names  row_count_est  row_count_err  distinct_count_est  distinct_count_err  null_count_est  null_count_err
        {o_d_id}      300000.00      1.00           10.00               1.00                0.00            1.00
        {o_id}        300000.00      1.00           2999.00             1.00                0.00            1.00
        {o_ol_cnt}    300000.00      1.00           11.00               1.00                0.00            1.00
        {o_w_id}      300000.00      1.00           10.00               1.00                0.00            1.00
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/938/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/memo/memo_test_/memo_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/memo/testdata/stats_quality/tpcc:1656:
         SELECT count(*)
        FROM
        (
            SELECT o_w_id, o_d_id, o_id
            FROM "order"
            WHERE o_carrier_id IS NULL
        )
        FULL OUTER JOIN
        (
            SELECT ol_w_id, ol_d_id, ol_o_id
            FROM order_line
            WHERE ol_delivery_d IS NULL
        )
        ON (ol_w_id = o_w_id AND ol_d_id = o_d_id AND ol_o_id = o_id)
        WHERE ol_o_id IS NULL OR o_id IS NULL
        output didn't match expected:
        @@ -6,17 +6,17 @@
          ├── key: ()
          ├── fd: ()-->(23)
          ├── select
          │    ├── save-table-name: consistency_12_select_2
          │    ├── columns: o_id:1(int) o_d_id:2(int) o_w_id:3(int) ol_o_id:11(int) ol_d_id:12(int) ol_w_id:13(int)
        - │    ├── stats: [rows=629393.8, distinct(1)=2999, null(1)=440513, distinct(2)=10, null(2)=440513, distinct(3)=10, null(3)=440513, distinct(11)=2999, null(11)=0, distinct(12)=10, null(12)=0, distinct(13)=10, null(13)=0]
        + │    ├── stats: [rows=629393.8, distinct(1)=2999, null(1)=440513, distinct(2)=10, null(2)=440513, distinct(3)=10, null(3)=440513, distinct(11)=2999, null(11)=-1.62589e-11, distinct(12)=10, null(12)=-1.62589e-11, distinct(13)=10, null(13)=-1.62589e-11]
          │    ├── full-join (merge)
          │    │    ├── save-table-name: consistency_12_merge_join_3
          │    │    ├── columns: o_id:1(int) o_d_id:2(int) o_w_id:3(int) ol_o_id:11(int) ol_d_id:12(int) ol_w_id:13(int)
          │    │    ├── left ordering: +13,+12,-11
          │    │    ├── right ordering: +3,+2,-1
        - │    │    ├── stats: [rows=899134, distinct(1)=2999, null(1)=629304, distinct(2)=10, null(2)=629304, distinct(3)=10, null(3)=629304, distinct(11)=2999, null(11)=0, distinct(12)=10, null(12)=0, distinct(13)=10, null(13)=0]
        + │    │    ├── stats: [rows=899134, distinct(1)=2999, null(1)=629304, distinct(2)=10, null(2)=629304, distinct(3)=10, null(3)=629304, distinct(11)=2999, null(11)=-2.3227e-11, distinct(12)=10, null(12)=-2.3227e-11, distinct(13)=10, null(13)=-2.3227e-11]
          │    │    ├── project
          │    │    │    ├── save-table-name: consistency_12_project_4
          │    │    │    ├── columns: ol_o_id:11(int!null) ol_d_id:12(int!null) ol_w_id:13(int!null)
          │    │    │    ├── stats: [rows=899134, distinct(11)=2999, null(11)=0, distinct(12)=10, null(12)=0, distinct(13)=10, null(13)=0]
          │    │    │    ├── ordering: +13,+12,-11
    --- FAIL: TestStatsQuality/tpcc (0.30s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatsQuality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/opt/memo.TestStatsQuality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12904062?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12904062?buildTab=artifacts#/) on master @ [5882e25f7efe24bcf7b8f0783170fd7c2f47eed1](https://github.com/cockroachdb/cockroach/commits/5882e25f7efe24bcf7b8f0783170fd7c2f47eed1):


```
                    }
                ],
                "histo_col_type": "VARCHAR(101)",
                "histo_version": 1,
                "name": "__auto__",
                "null_count": 0,
                "row_count": 10000
            }
        ]';
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12186/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/memo/memo_test_/memo_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/memo/testdata/stats_quality/tpch/q13:4:
        import [1 args]
        <no input to command>
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12186/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/memo/memo_test_/memo_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/memo/testdata/stats_quality/tpch/q13:18:
         SELECT
            c_count, count(*) AS custdist
        FROM (
            SELECT
                c_custkey,
                count(o_orderkey)
            FROM
                customer LEFT OUTER JOIN orders ON
                    c_custkey = o_custkey
                    AND o_comment NOT LIKE '%special%requests%'
            GROUP BY
                c_custkey
            ) AS c_orders (c_custkey, c_count)
        GROUP BY
            c_count
        ORDER BY
            custdist DESC,
            c_count DESC;
        output didn't match expected:
        @@ -20,11 +20,11 @@
               │    ├── key: (1)
               │    ├── fd: (1)-->(22)
               │    ├── right-join (hash)
               │    │    ├── save-table-name: q13_right_join_4
               │    │    ├── columns: c_custkey:1(int!null) o_orderkey:11(int) o_custkey:12(int) o_comment:19(varchar)
        -      │    │    ├── stats: [rows=503988.2, distinct(1)=148813, null(1)=0, distinct(11)=317522, null(11)=0, distinct(12)=99620.1, null(12)=0, distinct(19)=317522, null(19)=0]
        +      │    │    ├── stats: [rows=503988.2, distinct(1)=148813, null(1)=0, distinct(11)=317522, null(11)=1.10795e-11, distinct(12)=99620.1, null(12)=0, distinct(19)=317522, null(19)=1.10795e-11]
               │    │    ├── key: (1,11)
               │    │    ├── fd: (11)-->(12,19)
               │    │    ├── select
               │    │    │    ├── save-table-name: q13_select_5
               │    │    │    ├── columns: o_orderkey:11(int!null) o_custkey:12(int!null) o_comment:19(varchar!null)
        --- FAIL: TestStatsQuality/tpch/q13 (0.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatsQuality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -1652,24 +1652,25 @@ func (sb *statisticsBuilder) adjustNullCountsForOuterJoins(
 	// extension - such as right columns for non-matching rows in left joins.
 	switch joinType {
 	case opt.LeftJoinOp, opt.LeftJoinApplyOp:
-		if !rightColsAreEmpty {
-			colStat.NullCount += (rowCount - innerJoinRowCount) * leftNullCount / leftRowCount
+		if !rightColsAreEmpty && leftNullCount > 0 && rowCount > innerJoinRowCount {
+			addedRows := max(rowCount-innerJoinRowCount, epsilon)
+			colStat.NullCount += addedRows * leftNullCount / leftRowCount
 		}
 
 	case opt.RightJoinOp:
-		if !leftColsAreEmpty {
-			colStat.NullCount += (rowCount - innerJoinRowCount) * rightNullCount / rightRowCount
+		if !leftColsAreEmpty && rightNullCount > 0 && rowCount > innerJoinRowCount {
+			addedRows := max(rowCount-innerJoinRowCount, epsilon)
+			colStat.NullCount += addedRows * rightNullCount / rightRowCount
 		}
 
 	case opt.FullJoinOp:
-		leftJoinRowCount := max(innerJoinRowCount, leftRowCount)
-		rightJoinRowCount := max(innerJoinRowCount, rightRowCount)
-
-		if !leftColsAreEmpty {
-			colStat.NullCount += (rightJoinRowCount - innerJoinRowCount) * rightNullCount / rightRowCount
+		if !leftColsAreEmpty && rightNullCount > 0 && rightRowCount > innerJoinRowCount {
+			addedRows := max(rightRowCount-innerJoinRowCount, epsilon)
+			colStat.NullCount += addedRows * rightNullCount / rightRowCount
 		}
-		if !rightColsAreEmpty {
-			colStat.NullCount += (leftJoinRowCount - innerJoinRowCount) * leftNullCount / leftRowCount
+		if !rightColsAreEmpty && leftNullCount > 0 && leftRowCount > innerJoinRowCount {
+			addedRows := max(leftRowCount-innerJoinRowCount, epsilon)
+			colStat.NullCount += addedRows * leftNullCount / leftRowCount
 		}
 	}
 }
@@ -3006,15 +3007,6 @@ func (sb *statisticsBuilder) rowsProcessed(e RelExpr) float64 {
 		}
 		return e.Relational().Statistics().RowCount
 	}
-}
-
-// TODO(#115278): We should be able to replace this with Go's built-in max
-// function, but doing breaks some optimizer tests on ARM64.
-func max(a, b float64) float64 {
-	if a > b {
-		return a
-	}
-	return b
 }
 
 //////////////////////////////////////////////////

